File: remes/x/remes/keeper/shard_assignment.go
package keeper

import (
	"crypto/sha256"
	"fmt"

	sdk "github.com/cosmos/cosmos-sdk/types"
)

// CalculateShardID calculates a deterministic shard ID for a miner in a training round.
// Formula: (wallet_address + block_hash + round_id) % total_shards
// This ensures:
// - Same miner gets same shard within a round (stable)
// - Different miners get different shards (distributed)
// - Different rounds can have different assignments (flexible)
func (k Keeper) CalculateShardID(
	ctx sdk.Context,
	minerAddress string,
	trainingRoundID uint64,
	totalShards uint64,
) (uint64, error) {
	if totalShards == 0 {
		return 0, fmt.Errorf("total_shards cannot be zero")
	}

	// Get block hash (or block height as fallback)
	blockHash := ctx.HeaderHash()
	if len(blockHash) == 0 {
		// Fallback: use block height as deterministic seed
		blockHash = []byte(fmt.Sprintf("block_%d", ctx.BlockHeight()))
	}

	// Create deterministic input: miner_address + block_hash + round_id
	input := fmt.Sprintf("%s|%x|%d", minerAddress, blockHash, trainingRoundID)

	// Hash the input
	hash := sha256.Sum256([]byte(input))

	// Convert first 8 bytes to uint64
	var shardID uint64
	for i := 0; i < 8 && i < len(hash); i++ {
		shardID = shardID<<8 | uint64(hash[i])
	}

	// Modulo to get shard ID in range [0, totalShards)
	shardID = shardID % totalShards

	return shardID, nil
}

// VerifyShardAssignment verifies that a miner's shard assignment is correct.
func (k Keeper) VerifyShardAssignment(
	ctx sdk.Context,
	minerAddress string,
	claimedShardID uint64,
	trainingRoundID uint64,
	totalShards uint64,
) (bool, error) {
	expectedShardID, err := k.CalculateShardID(ctx, minerAddress, trainingRoundID, totalShards)
	if err != nil {
		return false, err
	}

	return claimedShardID == expectedShardID, nil
}

File: miner-engine/utils/shard_assignment.py
"""
Deterministic Shard Assignment for R3MES Training

Implements deterministic shard assignment using:
(wallet_address + block_hash + round_id) % total_shards

This ensures:
- Same miner gets same shard within a round (stable)
- Different miners get different shards (distributed)
- Different rounds can have different assignments (flexible)
"""

import hashlib
from typing import Optional


def calculate_shard_id(
    miner_address: str,
    block_hash: str,
    training_round_id: int,
    total_shards: int = 100,
) -> int:
    """
    Calculate deterministic shard ID for a miner in a training round.
    
    Args:
        miner_address: Miner's wallet address (bech32 format)
        block_hash: Block hash from blockchain (hex string)
        training_round_id: Current training round ID
        total_shards: Total number of shards (default: 100)
    
    Returns:
        Shard ID in range [0, total_shards)
    """
    if total_shards == 0:
        raise ValueError("total_shards cannot be zero")
    
    # Create deterministic input: miner_address + block_hash + round_id
    input_str = f"{miner_address}|{block_hash}|{training_round_id}"
    
    # Hash the input
    hash_bytes = hashlib.sha256(input_str.encode()).digest()
    
    # Convert first 8 bytes to integer
    shard_id = int.from_bytes(hash_bytes[:8], byteorder='big')
    
    # Modulo to get shard ID in range [0, total_shards)
    shard_id = shard_id % total_shards
    
    return shard_id


def verify_shard_assignment(
    miner_address: str,
    block_hash: str,
    training_round_id: int,
    claimed_shard_id: int,
    total_shards: int = 100,
) -> bool:
    """
    Verify that a miner's shard assignment is correct.
    
    Args:
        miner_address: Miner's wallet address
        block_hash: Block hash from blockchain
        training_round_id: Current training round ID
        claimed_shard_id: Shard ID claimed by miner
        total_shards: Total number of shards
    
    Returns:
        True if shard assignment is correct, False otherwise
    """
    expected_shard_id = calculate_shard_id(
        miner_address,
        block_hash,
        training_round_id,
        total_shards,
    )
    
    return claimed_shard_id == expected_shard_id

